### PR TITLE
jitter: add a new seed RNG "JITTER"

### PR DESCRIPTION
**Pull request by xnox**
This entropy source can be used instead of SEED-SRC. Sample openssl.cnf configuration is provided.

Note that rand-instances currently chain from public/private instances to primary, prior to consuming the seed. Thus currently a unique ESV needs to be obtained, and resue of jitterentropy.a certificate is not possible as is. Separately a patch will be sent to allow for unchaining public/private RAND instances for the purpose of reusing ESV.

Also I do wonder if it makes sense to create a fips variant of stock SEED-SRC entropy source, which in addition to using getrandom() also verifies that the kernel is operating in FIPS mode and thus is likely a validated entropy source. As in on Linux, check that /proc/sys/crypto/fips_enabled is set to 1, and similar checks on Windows / MacOS and so on.

Replaces:
- https://github.com/openssl/openssl/pull/24057

### CODE REVIEW

**Comment by nhorman** [comment]
you can ignore all the errors in the non-c files from the style check ci job, but the format errors in seed_src_jitter.c should likely be addressed.

**Comment by nhorman** [comment]
waiving style to ignore the files that we shouldn't really be checking (build.info, Configure, etc)

**Comment by xnox** [comment]
@nhorman i have fixed most style issues. Note most of this code was copied from legacy provider & seed-src. Remaining C code formatting issues are:

```
$ ./util/check-format-commit.sh HEAD
providers/jitterprov.c:118:more than one stmt:#define set_func(c, f) if (c == NULL) c = f; else if (c != f) return 0;
providers/jitterprov.c:118:code before 'else':#define set_func(c, f) if (c == NULL) c = f; else if (c != f) return 0;
providers/jitterprov.c:118:code after (expr):#define set_func(c, f) if (c == NULL) c = f; else if (c != f) return 0;
```

Which are the same as the ones in legacyprov.c, thus leaving as is.

**Comment by nhorman** [comment]
That's awesome, thank you for taking care of that!

File: providers/implementations/include/prov/names.h
Comment (paulidale) [inline providers/implementations/include/prov/names.h:314]: Is it worth considering just _JITTER_ for the name?

File: providers/implementations/rands/seed_src_jitter.c
Comment (paulidale) [inline providers/implementations/rands/seed_src_jitter.c:2]: The date is wrong :grin:

**Review by paulidale — COMMENTED** [review]
Look really good.

I'm not convinced about putting it in a separate provider but I'm not going to resist it.

The rate of change isn't a big deal IMO: an update to the FIPS provider's validation isn't going to require looking at the entropy source unless it has changed.

File: providers/jitterprov.c
Comment (paulidale) [inline providers/jitterprov.c:2]: Data again here.

**Comment by mattcaswell** [comment]
Cool! Nice work! I've not yet reviewed in detail but some immediate observations:

- This will require more documentation. In particular I might expect to see a doc/man7/EVP_RAND-JITTER-SEED-SRC.pod file, similar to the EVP_RAND-SEED-SRC.pod file that already exists. Since you are also adding a new provider I'd expect to see a doc/man7/OSSL_PROVIDER-jitter.pod file (similar to doc/man7/OSSL_PROVIDER-default.pod). Also, there are a number of places where we refer to the current OpenSSL supplied providers which might need some updates (e.g. ossl-guide-libraries-introduction.pod) - there are probably other places so we might need to do a quick scan of the docs to identify other locations like this.
- This is going to need some kind of tests. At least something simple to confirm that we can load the provider, set the seed src and use it. We might also want to consider if we can test this via CI.

I'm also interested in the design decision to put this in a separate provider. I assume this is because we want to be able to use this in conjunction with the fips provider. Did you consider putting it in the base and default providers?

**Comment by xnox** [comment]
@paulidale :
> I'm not convinced about putting it in a separate provider but I'm not going to resist it.

when I started I thought it was going to be big.... but it is tiny. Thus having it as a separate provider does seem like a gimmick. Many will not choose to build it, but those that will likely will want it in the base-and-default providers. Should I fold it into base-and-default providers like "SEED-SRC" already is?

**Comment by mattcaswell** [comment]
> Should I fold it into base-and-default providers like "SEED-SRC" already is?

This seems sensible to me in the absence of a good reason to have it in its own provider.

**Comment by xnox** [comment]
> I'm also interested in the design decision to put this in a separate provider. I assume this is because we want to be able to use this in conjunction with the fips provider.

Use in conjunction with the fips provider - yes. But also submit for certification separately / ahead of fips provider. As entropy ESV certificate needs to be obtained prior to submission for fips CMVP certificate.

> Did you consider putting it in the base and default providers?

I thought new provider would be better, but indeed having implemented all this having "JITTER" in the base and default providers would be less total compiled code and easier to use.

**Comment by xnox** [comment]
> * This is going to need some kind of tests. At least something simple to confirm that we can load the provider, set the seed src and use it. We might also want to consider if we can test this via CI.

There is a periodic build added in .github workflow. I think I can improve it by also compiling with rand-seed sources set to null, to make sure failures are detected correctly. Alternatively I have also been locally using systemd-run and making random syscall return an error (not abort, as glibc thread local storage uses getrandom) and then ensuring that rand calls are successful. If you want this as part of every presubmit the workflow job can be moved, but not sure it is that critical to check every time, instead of daily.

**Comment by xnox** [comment]
@nhorman 

In https://github.com/openssl/openssl/pull/24844#issuecomment-2221438898 you mentioned this:

> Looks like you've got a minor doc-nit to square up I'll start reviewing in the am

I missed that at the time, and not sure what that means. Were there documentation syntax or style errors somewhere at the time? I have now added more docs, but I don't know if I still have "a minor doc-nit" - can you please recheck and point it out to me, if still not fixed?

**Comment by xnox** [comment]
@mattcaswell I have now improved testing coverage, but please review if it is acceptable see https://github.com/openssl/openssl/pull/24844/commits/ef131f36afd88a9e49aaf6842805e6e051df5f79

File: providers/implementations/rands/seed_src_jitter.c
Comment (mattcaswell) [inline providers/implementations/rands/seed_src_jitter.c:218]: This seems high. But I see SEED-SRC has the same...

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: Configure
Comment (xnox) [reply Configure:1330]: @mattcaswell that was rejected in https://github.com/openssl/openssl/pull/24057#discussion_r1609037186 and requested to have jitter as a separate RNG class.

File: INSTALL.md
Comment (xnox) [reply INSTALL.md:537]: True

File: doc/man7/EVP_RAND-JITTER.pod
Comment (xnox) [reply doc/man7/EVP_RAND-JITTER.pod:12]: Ok. I was trying to mimick the documentation of the SEED-SRC which explains where the randomness comes from.

Comment (mattcaswell) [reply doc/man7/EVP_RAND-JITTER.pod:12]: I think it is sufficient to keep the second bit of this sentence (i.e. about it coming from CPU fluctuations)

File: providers/implementations/rands/seed_src_jitter.c
Comment (xnox) [reply providers/implementations/rands/seed_src_jitter.c:218]: @mattcaswell i'm not sure any of these parameters have any meaning. I chose to copy them as is, to maintain drop-in bug for bug compatibility with SEED-SRC. I hope we can consider separately if these should be removed/changed in jitter or seed-src or both separately from trying to add jitter. As I think this part of openssl is getting attention, for the first time in a little while.

File: crypto/rand/rand_lib.c
Comment (xnox) [reply crypto/rand/rand_lib.c:599]: @mattcaswell but i want to run all of them, and they all pass. Or like should this setting be made a define to allow compiling with `-DFALLBACK_SEED_SRC=JITTER` or some such? and when not defined, define it to "SEED-SRC"?

But also after implementing this, it does seem like a useful new configuration combo.

File: Configure
Comment (mattcaswell) [reply Configure:1330]: Ok - thanks for the reference. I see there was a lengthy discussion on this topic! I'll mark this as resolved.

File: providers/implementations/rands/seed_src_jitter.c
Comment (mattcaswell) [reply providers/implementations/rands/seed_src_jitter.c:218]: I'd be interested in @paulidale's view on this topic - but I don't consider this to be a blocker for this PR.

File: crypto/rand/rand_lib.c
Comment (mattcaswell) [reply crypto/rand/rand_lib.c:599]: > @mattcaswell but i want to run all of them, and they all pass.

Yeah, ok

> Or like should this setting be made a define to allow compiling with -DFALLBACK_SEED_SRC=JITTER or some such? and when not defined, define it to "SEED-SRC"?

That sounds quite nice actually.

File: providers/implementations/rands/seed_src_jitter.c
Comment (paulidale) [reply providers/implementations/rands/seed_src_jitter.c:218]: Way back I picked a number that was sufficiently high so that it would never be a problem in the future (famous last words).  Since this is a seed source, it can produce entropy to meet any desired level of strength so the specific number isn't important, it just has to be greater than or equal to the maximum strength of everything downstream.  There are a few places where this is used (DRBG parenting e.g.) but it's not widespread.

I don't think this should be addressed here.  Better would be an OTC discussion followed by a new PR if required.

Comment (xnox) [reply providers/implementations/rands/seed_src_jitter.c:218]: I agree with @paulidale above, that is my reading as well. With all the specs currently there are justifications of minimum entropy per byte, and minimum requirements to request at least N, where N so far has not been more than 256. And yes the current number is sufficiently high, but also a reasonable ceiling.

Marking this as resolved. If we need to schedule a followup about this, @mattcaswell please open an issue to track that.

Comment (mattcaswell) [reply providers/implementations/rands/seed_src_jitter.c:218]: I'm happy with the responses on this. No follow up needed.

File: doc/man7/EVP_RAND-JITTER.pod
Comment (xnox) [reply doc/man7/EVP_RAND-JITTER.pod:13]: @mattcaswell how about this?

```
This software seed source produces randomness based on tiny CPU "jitter" fluctuations.

It is available when OpenSSL is compiled with B<enable-jitter> option. When available it is listed in B<openssl list -random-generators> and B<openssl info -seeds>.
```

File: doc/man7/EVP_RAND-JITTER.pod
Comment (xnox) [reply doc/man7/EVP_RAND-JITTER.pod:72]: I copied things as is from the SEED-SRC usage.

Adding example that uses jitter to the set_seed_source_type manpage.

File: providers/implementations/rands/seed_src_jitter.c
Comment (xnox) [reply providers/implementations/rands/seed_src_jitter.c:106]: yes, need to do this.

File: crypto/info.c
Comment (paulidale) [inline crypto/info.c:194]: style nit: blank line betwixt declarations and code.

File: Configure
Comment (paulidale) [inline Configure:1329]: _As an alternative_ perhaps?

File: Configure
Comment (paulidale) [inline Configure:1330]: consider: _Or specify alternate_

File: doc/man7/EVP_RAND.pod
Comment (paulidale) [inline doc/man7/EVP_RAND.pod:230]: can specify **a** third-party

**Review by paulidale — APPROVED** [review]
Wording suggestions, approved with or without.

File: providers/implementations/rands/seed_src_jitter.c
Comment (nhorman) [inline providers/implementations/rands/seed_src_jitter.c:95]: can you elaborate on the reasoning for this comment?

Comment (xnox) [reply providers/implementations/rands/seed_src_jitter.c:95]: @nhorman _safe API automatically increases OCR and ignores intermittent and permanent failures. Security policies for all publicly available jitter entropy sources thus do not permit using _safe() api call.

> the jent_read_entropy_safe of the master branch still cannot be used because it re-allocates the entropy collector if a health test failure is observed. Preventing re-init and preventing the oversampling rate from increasing would make the function FIPS compliant. I don't think that the function will be re-implemented upstream to make it FIPS compliant, because there is another function that is already FIPS compliant: jent_read_entropy. You should use jent_read_entropy.

**Review by paulidale — APPROVED** [review]

**Comment by xnox** [comment]
@mattcaswell I believe I have addressed all the raised issues; but please let me know if anything is missing or if further improvements are desired.

**Comment by paulidale** [comment]
Ping @openssl/committers for second review.

**Comment by slontis** [comment]
Please add a CHANGES entry for this.
Maybe it should state that the jitter library is not distributed with openssl?

**Comment by paulidale** [comment]
There is a changes entry present already.
The install instructions aren't explicit about the library not being included but do hints at it.

File: providers/implementations/rands/seed_src_jitter.c
Comment (slontis) [inline providers/implementations/rands/seed_src_jitter.c:91]: Reading Section 3.5 of http://chronox.de/jent/releases/3.5.0/CPU-Jitter-NPTRNG-v3.4.0.pdf..
Is this collector_alloc supposed to be done on every call?
It also looks like it is supposed to grabbed once and the result not used (@paulidale is this done in a layer above?)

**Comment by xnox** [comment]
> Should the CHANGES entry state that the jitter library is not distributed with openssl?

Define distributed :-) as build scripts explicitly statically link it in, but indeed it is expected to be compiled and provided separately.

Note that jitterentropy library has separate strict build instructions, and must be compiled unoptimised. Thus care needs to be taken how it is compiled.

File: providers/implementations/rands/seed_src_jitter.c
Comment (xnox) [reply providers/implementations/rands/seed_src_jitter.c:91]: One can do as one wishes. Most other users of this API create fresh collector on each use. See other users of these APIs in jitter rng (same author) and aws-lc. I think libgcrypt has jitter entropy library too - but didn't check how it is used there yet.

Note this is closer aligned with the rare failure conditions, as upon failure no output from the given collector should be used again nor will it produce any.

I have not done comparative performance benchmarking of the two approaches... As I am not sure what constitutes performance here.

Comment (paulidale) [reply providers/implementations/rands/seed_src_jitter.c:91]: Separation by creating new instances each call has merit I think.  It provides some measure of independence for each generation which is good.  The downside will be a performance hit, but I doubt it's large compared to the cost of the generation (it's essentially one additional iteration of the collector).

**Review by slontis — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
Merged, thanks very much for this contribution.

**Comment by t8m** [comment]
Please see https://github.com/openssl/openssl/actions/runs/10174927934 @xnox will you please create a PR to fix it ASAP?

**Comment by xnox** [comment]
> Please see https://github.com/openssl/openssl/actions/runs/10174927934 @xnox will you please create a PR to fix it ASAP?

Sure will look.

**Pull request closed without merging**
